File: lox.net/Scanner.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace lox.net
{
    class Scanner
    {
        private readonly string source;

        private int start = 0;
        private int current = 0;

        private int line = 1;

        //private readonly List<Token> tokens = new List<Token>();

        public Scanner(string source)
        {
            this.source = source;
        }

        public async IAsyncEnumerable<Token> ScanTokensAsync()
        {
            while (!IsAtEnd())
            {
                // We are at the beginning of the next lexeme.
                start = current;
                var token = await ScanTokenAsync();
                if (token != null)
                {
                    yield return token;
                }
                // Rlse, we got an error but keep on scanning.
            }

            yield return new Token(TokenType.EOF, "", null, line);
        }

        private bool IsAtEnd() =>
            current >= source.Length;

        private async Task<Token?> ScanTokenAsync()
        {
            char c = Advance();
            switch (c)
            {
                case '(':
                    return CreateToken(TokenType.LEFT_PAREN);
                case ')':
                    return CreateToken(TokenType.RIGHT_PAREN);
                case '{':
                    return CreateToken(TokenType.LEFT_BRACE);
                case '}':
                    return CreateToken(TokenType.RIGHT_BRACE);
                case ',':
                    return CreateToken(TokenType.COMMA);
                case '.':
                    return CreateToken(TokenType.DOT);
                case '-':
                    return CreateToken(TokenType.MINUS);
                case '+':
                    return CreateToken(TokenType.PLUS);
                case ';':
                    return CreateToken(TokenType.SEMICOLON);
                case '/':
                    // TODO: Need to hadle // comments.
                    return CreateToken(TokenType.SLASH);
                case '*':
                    return CreateToken(TokenType.STAR);
                case '!':
                    return CreateToken(Match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                case '=':
                    return CreateToken(Match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                case '<':
                    return CreateToken(Match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                case '>':
                    return CreateToken(Match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                default:
                    // TODO: Report the actual unexpected character.
                    // TODO: Group together a run of unexpected characrters in a single error.
                    await Program.ErrorAsync(line, "Unexpected character.");
                    return null;
            }
        }

        private char Advance()
        {
            current++;
            return source[current - 1];
        }

        private Token CreateToken(TokenType type)
        {
            return CreateToken(type, null);
        }

        private Token CreateToken(TokenType type, object? literal)
        {
            var text = source.Substring(start, current - start);
            return new Token(type, text, literal, line);
        }

        private bool Match(char expected)
        {
            if (IsAtEnd())
            {
                return false;
            }
            
            if (source[current] != expected)
            {
                return false;
            }

            current++;
            return true;
        }
    }
}

File: kotlin/klox/src/main/kotlin/eu/grigoriu/craftinginterpreters/klox/Scanner.kt
package eu.grigoriu.craftinginterpreters.klox

class Scanner(
    private val source: String,
    private val errorReporter: ErrorReporter
) {
    companion object {
        private val keywords = mapOf(
            "and"    to TokenType.AND,
            "class"  to TokenType.CLASS,
            "else"   to TokenType.ELSE,
            "false"  to TokenType.FALSE,
            "for"    to TokenType.FOR,
            "fun"    to TokenType.FUN,
            "if"     to TokenType.IF,
            "nil"    to TokenType.NIL,
            "or"     to TokenType.OR,
            "print"  to TokenType.PRINT,
            "return" to TokenType.RETURN,
            "super"  to TokenType.SUPER,
            "this"   to TokenType.THIS,
            "true"   to TokenType.TRUE,
            "var"    to TokenType.VAR,
            "while"  to TokenType.WHILE
        )
    }

    private val tokens = mutableListOf<Token>()

    private var start = 0
    private var current = 0
    private var line = 1

    fun scanTokens(): List<Token> {
        while (!isAtEnd()) {
            start = current
            scanToken()
        }

        tokens.add(Token(TokenType.EOF, "", null, line))
        return tokens
    }

    private fun scanToken() {
        when (val c = advance()) {
            '(' -> {
                addToken(TokenType.LEFT_PAREN)
            }
            ')' -> {
                addToken(TokenType.RIGHT_PAREN)
            }
            '{' -> {
                addToken(TokenType.LEFT_BRACE)
            }
            '}' -> {
                addToken(TokenType.RIGHT_BRACE)
            }
            ',' -> {
                addToken(TokenType.COMMA)
            }
            '.' -> {
                addToken(TokenType.DOT)
            }
            '-' -> {
                addToken(TokenType.MINUS)
            }
            '+' -> {
                addToken(TokenType.PLUS)
            }
            ';' -> {
                addToken(TokenType.SEMICOLON)
            }
            '*' -> {
                addToken(TokenType.STAR)
            }
            '!' -> {
                addToken(if (match('=')) TokenType.BANG_EQUAL else TokenType.BANG)
            }
            '=' -> {
                addToken(if (match('=')) TokenType.EQUAL_EQUAL else TokenType.EQUAL)
            }
            '<' -> {
                addToken(if (match('=')) TokenType.LESS_EQUAL else TokenType.LESS)
            }
            '>' -> {
                addToken(if (match('=')) TokenType.GREATER_EQUAL else TokenType.GREATER)
            }
            '/' -> {
                if (match('/')) {
                    // A comment goes until the end of the line.
                    while (peek() != '\n' && !isAtEnd()) {
                        advance()
                    }
                } else {
                    addToken(TokenType.SLASH)
                }
            }
            // Ignore whitespace.
            ' ', '\r', '\t' -> {}
            '\n' -> {
                line++
            }
            '"' -> {
                string()
            }
            else -> {
                if (isDigit(c)) {
                    number()
                } else if (isAlpha(c)) {
                    identifier()
                } else {
                    errorReporter.error(line, "Unexpected character: $c")
                }
            }
        }
    }

    private fun advance(): Char {
        return source[current++]
    }

    private fun addToken(tokenType: TokenType, literal: Any? = null) {
        val text = source.substring(start, current)
        tokens.add(Token(tokenType, text, literal, line))
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd()) {
            return false
        }
        if (source[current] != expected) {
            return false
        }

        current++
        return true
    }

    private fun peek(): Char {
        return if (isAtEnd()) 0.toChar() else source[current]
    }

    private fun peekNext(): Char {
        if (current + 1 >= source.length) {
            return 0.toChar()
        }
        return source[current + 1]
    }

    private fun isAtEnd(): Boolean {
        return current >= source.length
    }

    private fun string() {
        while (peek() != '"' && !isAtEnd()) {
            // Count lines inside of strings.
            if (peek() == '\n') {
                line++
            }
            advance()
        }

        if (isAtEnd()) {
            errorReporter.error(line, "Unterminated string.")
            return
        }

        // The closing ".
        advance()

        // Trim the surrounding quotes.
        val value = source.substring(start + 1, current - 1)
        addToken(TokenType.STRING, value)
    }

    private fun isDigit(c: Char): Boolean {
        return c in '0'..'9'
    }

    private fun isAlpha(c: Char): Boolean {
        return (c in 'a'..'z') || (c in 'A'..'Z') || c == '_'
    }

    private fun isAlphaNumeric(c: Char): Boolean {
        return isAlpha(c) || isDigit(c)
    }

    private fun number() {
        while (isDigit(peek())) {
            advance()
        }

        // Look for a fractional part.
        if (peek() == '.' && isDigit(peekNext())) {
            // Consume the ".".
            advance()

            while (isDigit(peek())) {
                advance()
            }
        }

        addToken(TokenType.NUMBER, source.substring(start, current).toDouble())
    }

    private fun identifier() {
        while (isAlphaNumeric(peek())) {
            advance()
        }

        val text = source.substring(start, current)
        val tokenType = keywords[text] ?: TokenType.IDENTIFIER

        addToken(tokenType)
    }
}